url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/graph/GraphLayoutCache.java

```diff
@@ -0,0 +1,2953 @@
+			return previousAttributes;
+		}
+
+		public Rectangle2D getDirtyRegion() {
+			return dirtyRegion;
+		}
+
+		public void setDirtyRegion(Rectangle2D dirty) {
+			this.dirtyRegion = dirty;
+			
+		}
+		
+		/**
+		 * Redoes a change.
+		 * 
+		 * @exception CannotRedoException
+		 *                if the change cannot be redone
+		 */
+		public void redo() throws CannotRedoException {
+			super.redo();
+			execute();
+		}
+
+		/**
+		 * Undoes a change.
+		 * 
+		 * @exception CannotUndoException
+		 *                if the change cannot be undone
+		 */
+		public void undo() throws CannotUndoException {
+			super.undo();
+			execute();
+		}
+
+		/**
+		 * Execute this edit such that the next invocation to this method will
+		 * invert the last execution.
+		 */
+		public void execute() {
+			GraphModel model = getModel();
+			changedCells.clear();
+			// Remember or restore hidden cells
+			if (hidden != null)
+				for (int i = 0; i < hidden.length; i++)
+					if (hidden[i] != null)
+						mapping.put(hidden[i].getCell(), hidden[i]);
+			
+			if (invisible != null && invisible.length >0) {
+				CellView[] invisibleViews = new CellView[invisible.length];
+				invisibleViews = getMapping(invisible, true);
+				Rectangle2D changedBounds = getBounds(invisibleViews);
+				dirtyRegion = RectUtils.union(dirtyRegion, changedBounds);
+			}
+
+						
+			if (!remembersCellViews) // already remembered
+				hidden = getMapping(invisible);
+			// Handle visibility
+			boolean updatePorts = setVisibleImpl(visible, true)
+					| setVisibleImpl(invisible, false);
+			if (visible != null) {
+				for (int i = 0; i < visible.length; i++) {
+					changedCells.add(visible[i]);
+
+					// Only calls if not inserted
+					if (cells == null)
+						cellExpanded(visible[i]);
+				}
+			}
+			if (invisible != null)
+				for (int i = 0; i < invisible.length; i++)
+					changedCells.add(invisible[i]);
+			// Swap arrays
+			Object[] tmp = visible;
+			visible = invisible;
+			invisible = tmp;
+			// Handle attributes
+			if (attributes != null) {
+				previousAttributes = attributes;
+				changedCells.addAll(attributes.keySet());
+			}
+			if (updatePorts)
+				updatePorts();
+			// Add ancestor cells to changed cells
+			Set parentSet = new HashSet();
+			Iterator it = changedCells.iterator();
+			while (it.hasNext()) {
+				Object parent = model.getParent(it.next());
+				while (parent != null) {
+					parentSet.add(parent);
+					parent = model.getParent(parent);
+				}
+			}
+			changedCells.addAll(parentSet);
+			Set ctx = DefaultGraphModel.getEdges(getModel(), changedCells
+					.toArray());
+			context = getMapping(ctx.toArray());
+			Set allChangedCells = new HashSet(changedCells);
+			allChangedCells.addAll(ctx);
+			CellView[] allChangedCellViews = getMapping(allChangedCells.toArray());
+			Rectangle2D changedBounds = getBounds(allChangedCellViews);
+			dirtyRegion = RectUtils.union(dirtyRegion, changedBounds);
+			if (attributes != null) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It seems unusual to set the value of a variable only when it is already set. You might want to switch the null check of != on line 2321 to == check before setting the field value on line 2322 or verify if the check is needed at all.